compile/constant_value.go

```diff
@@ -179,9 +319,12 @@ type ConstReference struct {
 }
 
 // Link for ConstReference.
-func (c ConstReference) Link(scope Scope) (ConstantValue, error) {
-	// The reference has already been verified. Nothing to do.
-	return c, nil
+func (c ConstReference) Link(scope Scope, t TypeSpec) (ConstantValue, error) {
+	cv, err := typeEqualsOrCastError(c, c.Target.Type, t)
```

Comment:
As discussed offline: If the types match, we'll use the constant reference as-is (so the generated code will have a reference to a declared constant), otherwise, we'll resolve the reference away to a new constant value.